components/brave_rewards/browser/rewards_protocol_navigation_throttle.cc

```diff
@@ -68,10 +92,17 @@ bool IsValidWalletProviderRedirect(
   return true;
 }
 
-void LoadRewardsURL(const GURL& redirect_url,
-                    content::WebContents::OnceGetter web_contents_getter,
-                    ui::PageTransition page_transition) {
-  content::WebContents* web_contents = std::move(web_contents_getter).Run();
+GURL TransformUrl(const GURL& url) {
+  DCHECK(url.is_valid());
+
+  return GURL(base::StrCat(
+      {"chrome://rewards/",
+       base::TrimString(url.path(), "/", base::TrimPositions::TRIM_LEADING),
```

Comment:
it's not clear to me exactly what's going on here, but it doesn't seem safe to just pass through the path like this. What are the expected values? cc @diracdeltas

Reply:
looks like there is something in `IsValidWalletProviderRedirect`, but it seems insufficient to me. What if they did something like `rewards://uphold/../settings/.....`?

Reply:
talked to brian on slack and basically we are not sure if `IsValidWalletProviderRedirect` can be bypassed if the redirect_url is something like `rewards://uphold/../settings/` because `brave://rewards/uphold/../settings` is resolved to `brave://rewards/settings`. if it is bypassable, then a wallet provider can redirect to any page in brave://rewards, not just under their own path. cc @szilardszaloki @deeppandya can we add a test for this

Reply:
There's an easy fix for this: `GURL`'s constructor already does the normalization, so we just have to make sure we validate the redirect _after_ transforming the redirect URL. For instance, if bitFlyer wanted to redirect to `rewards://bitflyer/../uphold/authorization`, it would work with the current code, but not if we validated the redirect after the URL transform (which yields `chrome://rewards/uphold/authorization`).

Reply:
@szilardszaloki according to @deeppandya GURL does not automatically normalize the url. From @deeppandya ` then i tried `GURL("rewards://uphold/../settings/").path()` which shows //uphold/../settings/`

Reply:
@bridiver Most probably because `rewards://` is not registered as standard scheme. Works fine with `chrome://`/`brave://`.

Reply:
@szilardszaloki that is possible, I haven't actually checked myself. If that's the case then we can do the check after Transform assuming the `..` is gone at that point

Reply:
That's what my fix does above. 🙂

Reply:
i didn't know about registered as standard scheme part.

Reply:
Resolving this — fixed by the patch [here](https://github.com/brave/brave-core/pull/20332#discussion_r1355270205) (which is already applied).